drivers/periph_common/hwrng.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2019 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_periph_hwrng
+ * @{
+ *
+ * @file        hwrng.c
+ * @brief       Copy random values to buffer
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include "periph_cpu.h"
+#include "periph/hwrng.h"
+
+#if defined(MODULE_PERIPH_HWRNG) && HWRNG_HAS_GENERIC_READ
+
+void hwrng_read(void *buf, unsigned int num)
+{
+    uint32_t *b32 = (uint32_t *)buf;
```

Comment:
This is undefined behavior, as `void *buf` has no alignment requirements, but `uint32_t *` has an alignment requirement of four bytes.

Lets assume the compiler would be generating code for an ARMv5 platform. Because of `uint32_t *b32` has an alignment requirement of four, the compiler is free to use the `STR` instruction to store four bytes at a time. On ARMv5 however, the two least significant bits of the address are ignored (rounding down to the next multiple of four). That could go horribly wrong if `buf` points to a char array, that happens to be allocated on an odd address.

For platforms that do support unalignment memory accesses (like more recent ARMs or x86), those mostly come with a significant performance penalty.

To get this code safe, you would have to copy byte-wise until you reach an address aligned to four bytes. From that point on you could copy four bytes at a time.

I bet however, that this is not worth the effort: The simple (and safe) implementation to just copy byte-wise is something compilers often face - I bet they will be able to optimize loop quite well.

Reply:
With that it's down to `4464 KiB/s` in `tests/rng` now, but this is exercising the worst case where it will call
```C
uint32_t result;
hwrng_read(&result, 4);
```
in a loop.

Still an improvement though.